pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -0,0 +1,867 @@
+		secretNames = append(secretNames, secretName)
+	}
+	sort.Strings(secretNames)
+
+	var configs []*tls.CertAndStores
+	for _, secretName := range secretNames {
+		configs = append(configs, tlsConfigs[secretName])
+	}
+
+	return configs
+}
+
+func getCertificateBlocks(secret *corev1.Secret, namespace, secretName string) (string, string, error) {
+	var missingEntries []string
+
+	tlsCrtData, tlsCrtExists := secret.Data["tls.crt"]
+	if !tlsCrtExists {
+		missingEntries = append(missingEntries, "tls.crt")
+	}
+
+	tlsKeyData, tlsKeyExists := secret.Data["tls.key"]
+	if !tlsKeyExists {
+		missingEntries = append(missingEntries, "tls.key")
+	}
+
+	if len(missingEntries) > 0 {
+		return "", "", fmt.Errorf("secret %s/%s is missing the following TLS data entries: %s",
+			namespace, secretName, strings.Join(missingEntries, ", "))
+	}
+
+	cert := string(tlsCrtData)
+	if cert == "" {
+		missingEntries = append(missingEntries, "tls.crt")
+	}
+
+	key := string(tlsKeyData)
+	if key == "" {
+		missingEntries = append(missingEntries, "tls.key")
+	}
+
+	if len(missingEntries) > 0 {
+		return "", "", fmt.Errorf("secret %s/%s contains the following empty TLS data entries: %s",
+			namespace, secretName, strings.Join(missingEntries, ", "))
+	}
+
+	return cert, key, nil
+}
+
+// loadServices is generating a WRR service, even when there is only one target.
+func loadServices(client Client, namespace string, targets []v1alpha1.HTTPRouteForwardTo) (*dynamic.Service, map[string]*dynamic.Service, error) {
+	services := map[string]*dynamic.Service{}
+
+	wrrSvc := &dynamic.Service{
+		Weighted: &dynamic.WeightedRoundRobin{
+			Services: []dynamic.WRRService{},
+		},
+	}
+
+	for _, forwardTo := range targets {
+		svc := dynamic.Service{
+			LoadBalancer: &dynamic.ServersLoadBalancer{
+				PassHostHeader: func(v bool) *bool { return &v }(true),
+			},
+		}
+
+		// TODO Handle BackendRefs
+
+		if forwardTo.ServiceName != nil {
+			service, exists, err := client.GetService(namespace, *forwardTo.ServiceName)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if !exists {
+				return nil, nil, errors.New("service not found")
+			}
+
+			if len(service.Spec.Ports) > 1 && forwardTo.Port == 0 {
+				// If the port is unspecified and the backend is a Service
+				// object consisting of multiple port definitions, the route
+				// must be dropped from the Gateway. The controller should
+				// raise the "ResolvedRefs" condition on the Gateway with the
+				// "DroppedRoutes" reason.  The gateway status for this route
+				// should be updated with a condition that describes the error
+				// more specifically.
+				log.WithoutContext().Errorf("A multiple ports Kubernetes Service cannot be used if unspecified forwardTo.Port")
+				continue
+			}
+
+			var portName string
+			var portSpec corev1.ServicePort
+			var match bool
+
+			for _, p := range service.Spec.Ports {
+				if forwardTo.Port == 0 || p.TargetPort.IntVal == int32(forwardTo.Port) {
```

Comment:
I think we should also check that the `TargetPort` is of `int` type, because the `TargetPort` of a service can reference a container port by name.